polkadot/node/network/collator-protocol/src/validator_side_experimental/state.rs

```diff
@@ -14,19 +14,550 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
-use crate::validator_side_experimental::{peer_manager::Backend, Metrics, PeerManager};
-use sp_keystore::KeystorePtr;
+use crate::{
+	validator_side::notify_collation_seconded,
+	validator_side_experimental::{
+		collation_manager::CollationManager,
+		common::{
+			Advertisement, CanSecond, CollationFetchResponse, PeerInfo, PeerState,
+			ProspectiveCandidate, TryAcceptOutcome, INVALID_COLLATION_SLASH,
+		},
+		error::{Error, FatalResult},
+		peer_manager::Backend,
+		Metrics, PeerManager,
+	},
+	LOG_TARGET,
+};
+use fatality::Split;
+use futures::stream::FusedStream;
+use polkadot_node_network_protocol::{peer_set::CollationVersion, OurView, PeerId};
+use polkadot_node_primitives::{SignedFullStatement, Statement};
+use polkadot_node_subsystem::{
+	messages::{CandidateBackingMessage, IfDisconnected, NetworkBridgeTxMessage},
+	CollatorProtocolSenderTrait,
+};
+use polkadot_node_subsystem_util::{request_session_index_for_child, runtime::recv_runtime};
+use polkadot_primitives::{
+	BlockNumber, CandidateReceiptV2 as CandidateReceipt, Hash, Id as ParaId,
+};
+use std::time::Duration;
 
 /// All state relevant for the validator side of the protocol lives here.
 pub struct State<B> {
 	peer_manager: PeerManager<B>,
-	keystore: KeystorePtr,
-	metrics: Metrics,
+	collation_manager: CollationManager,
+	_metrics: Metrics,
 }
 
 impl<B: Backend> State<B> {
 	/// Instantiate a new subsystem `State`.
-	pub fn new(peer_manager: PeerManager<B>, keystore: KeystorePtr, metrics: Metrics) -> Self {
-		Self { peer_manager, keystore, metrics }
+	pub fn new(
+		peer_manager: PeerManager<B>,
+		collation_manager: CollationManager,
+		metrics: Metrics,
+	) -> Self {
+		Self { peer_manager, collation_manager, _metrics: metrics }
+	}
+
+	/// Handle a new peer connection.
+	pub async fn handle_peer_connected<Sender: CollatorProtocolSenderTrait>(
+		&mut self,
+		sender: &mut Sender,
+		peer_id: PeerId,
+		version: CollationVersion,
+	) {
+		let outcome = self
+			.peer_manager
+			.try_accept_connection(
+				sender,
+				peer_id,
+				PeerInfo { version, state: PeerState::Connected },
```

Comment:
Passing PeerState here seems a bit redundant?

Reply:
I see that this is not used anywhere else, so we could hardcode it to  `PeerState::Connected` inside `try_accept_connection()` for the moment. But I don't know. Personally I would prefer to let `try_accept_connection()` receive the whole `PeerInfo` since it's more generic this way.

I don't know if I'm missing something. I can change it if there are strong preferences related to this.